### PR TITLE
support checking memory locks at read index

### PR DESCRIPTION
**Pull request by sticnarf**
With async commit, locks should be visible to readers before they are written to the raftstore. This is also true for replica reads.

The unapplied locks are stored in the memory of the leader. To make followers or learners able to see the unapplied locks, necessary information should be sent to the leader in `ReadIndex` to check if any lock in memory.

### CODE REVIEW

**Review by 5kbpers — APPROVED** [review]

**Review by youjiali1995 — APPROVED** [review]

File: proto/kvrpcpb.proto [hunk 1]
```diff
@@ -561,11 +561,19 @@ message SplitRegionResponse {
 // Sent from TiFlash to a TiKV node.
 message ReadIndexRequest{
     Context context = 1;
+
+    // TiKV checks the given range if there is any unapplied lock
+    // blocking the read request.
+    uint64 start_ts = 2;
+    repeated KeyRange ranges = 3;
```
Comment (solotzg) [inline proto/kvrpcpb.proto:568]: What if `ranges` is empty ? Does it mean full range of this region ?

Comment (sticnarf) [reply proto/kvrpcpb.proto:568]: If `ranges` is empty, no range will be checked. The start and end bound of the region should be passed if ithe full range needs to be checked.

**Comment by MyonKeminta** [comment]
Is there any other problem with this PR? If no I think it can be merged

**Comment by sticnarf** [comment]
> Is there any other problem with this PR? If no I think it can be merged

I think https://github.com/tikv/tikv/pull/8669 should be reviewed first. If there's no problem for that PR then we can merge it.

**Pull request merged**